model/map/src/main/java/org/keycloak/models/map/group/MapGroupAdapter.java

```diff
@@ -101,8 +102,8 @@ public String getParentId() {
 
     @Override
     public Stream<GroupModel> getSubGroupsStream() {
-        return session.groups().getGroupsStream(realm)
-                .filter(groupModel -> getId().equals(groupModel.getParentId()));
+        // TODO: make this a proper interface on GroupProvider
+        return ((MapGroupProvider) session.getProvider(GroupProvider.class)).getGroupsByParentId(realm, this.getId());
```

Comment:
We need to think of other approach due to following constraints:
- System-wide group provider might not be a map provider originating this group.

Do you recall the trick with anonymous adapter below? Perhaps would be useful in here.
https://github.com/keycloak/keycloak/blob/44cd6cd5fb6811853405e9ba3b6d1624dc280f2e/model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java#L61-L66

Something like this in the `MapGroupProvider`:
```java
    private Function<MapGroupEntity, GroupModel> entityToAdapterFunc(RealmModel realm) {
        // Clone entity before returning back, to avoid giving away a reference to the live object to the caller
        return origEntity -> new MapGroupAdapter(session, realm, origEntity) {
            public Stream<GroupModel> getSubGroupsStream() {
                return getGroupsByParentId(realm, this.getId());
            }
        };
    }
```

Reply:
This would work. However, I am not quite following what you mean by `System-wide group provider might not be a map provider originating this group.` What other provider could be originating this group?

Reply:
Any group provider. We cannot anticipate all the future uses.

Taking an example from current state - today there is `RealmCacheSession` and `JpaRealmProvider` implementing group provider. Could a `GroupAdapter`s originating from `JpaRealmProvider` rely on the fact that its originator is `session.groups()`?

Still in this case, we need to redirect the call to `getSubGroupsStream()` to the same provider's `getGroupsByParentId` - and the only way I can see at this moment is to override it in the anonymous class created in the provider itself.

Reply:
Given other implementations of `GroupProvider` running alongside `MapGroupProvider`, are we sure that only `MapGroupProvider` originating this entity contains its children? 

If we count on this scenario in the future, shouldn't we use `session.groups()` everywhere rather than searching only within `MapGroupProvider`?